### ISSUE TITLE
Move Winit integrations to a `window` feature 

### ISSUE THREAD

**Issue by AlexiWolf** [AlexiWolf/wolf_engine#26]
Move the Winit integrations provided by PR #25 behind a `window` feature, and revert the core API back to the simple
main-loop it was using previously.

### Goals

- Split Winit integrations from the core code by adding a `window` feature.
- Switch to the idea of a minimal "core" API on top of which features add new functionality.

### Non-Goals

- Provide any new behaviors.  This feature is *ONLY* moving Winit integration behind a feature, and cleaning up the
  mess from PR #25.

### Context and Purpose

The Winit integration provided by PR #25 replaced Wolf Engine's main loop with an `EventLoop` provided by Winit.  The
integration works perfectly well and should be kept as-is, but trying to include it as part of the core engine is
causing several issues:

- The main issue is the engine will no longer run in a headless server environment because Winit expects there to be a
  windowing system installed.  Trying to start the engine will result in a panic.

- Additionally, the dependency on an `EventLoop` causes issues with existing tests because, for platform compatibility
  reasons, the `EventLoop` does not allow initialization off the main thread.  Because tests are run in parallel, each
  on their own thread, this causes all the tests which need the engine or need to create a `Context` object to fail.
  The current work-around of providing separate methods to create a `Context` object without an event loop is not an
  acceptable solution because it's messy and confusing.

- Users are forced to compile and include Winit in their codebase whether or not they actually want to use it.

I want to revert back to using a simple main-loop, exactly as we were doing before PR #25, and only include integration
with Winit when the `window` feature is enabled.  This solution will help with all of the issues described above.

### Open Questions

Things we are unsure about.  This is an open invitation for new questions and answers.

- Q: How exactly should we add the Winit features on top of the core?
  - A: It has been decided that features just enable additional bits of code along side what's already there.  This 
    feature is no different.

# Overview

The window feature will enable a few changes to the core setup.  The most notable changes are additions to the `ContextBuilder` object, and the `WolfEngine`

**NOTE:** All changes are walled behind the "window" feature flag.  No changes are made to the core API.

### Engine Changes

- Add an additional `run_with_event_loop` method which takes a Winit `EventLoop` and uses it to run the engine.
- Revert the `run` method to only take a `State` and run in a simple loop, as it was before #25.

### Context Builder Changes

- Add an optional `EventLoop` field.
- The `new` method does not create an `EventLoop`, and it initializes the `event_loop` field to `None`.
-  Add a `with_create_event_loop` method that creates a new `EventLoop`.
- Add a `with_event_loop` method which creates the `ContextBuilder` using a pre-existing `EventLoop`.
- Add a `build_with_event_loop` method with one that also returns the `EventLoop` alongside the `Context` object.
- Revert the `build` method will only build the `Context` object, as it did before #25.
